### PR TITLE
Improve formatting of React `useImperativeHandle` hook

### PR DESCRIPTION
**Pull request by Jaswanth-Sriram-Veturi**
## Description

Improve formatting of React useImperativeHandle hook
Closes: #16033 

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/print/call-arguments.js [hunk 1]
```diff
@@ -318,12 +328,19 @@ function isHopefullyShortCallArgument(node) {
 
 function isReactHookCallWithDepsArray(args) {
```
Comment (sosukesuzuki) [inline src/language-js/print/call-arguments.js:329]: What do you think about following refactoring?

```js
/**
 * Checks if the arguments of a function are a call to a React Hook with a dependencies array.
 */
function isReactHookCallWithDepsArray(args) {
  if (args.length === 2) {
    /**
     * useEffect(() => {
     *   // do something
     * }, [dep1, dep2, dep2])
     */
    return isValidHookCallbackAndDepsFormat(args, /* baseIndex */ 0);
  }
  if (args.length === 3) {
    /**
     * useImperativeHandle(ref, () => {
     *   // do something
     * }, [dep1, dep2, dep2]);
     */
    return (
      args[0].type === "Identifier" &&
      isValidHookCallbackAndDepsFormat(args, /* baseIndex */ 1)
    );
  }
  return false;
}

function isValidHookCallbackAndDepsFormat(args, baseIndex) {
  const maybeArrowFunction = args[baseIndex];
  const maybeDepsArrary = args[baseIndex + 1];
  return (
    maybeArrowFunction.type === "ArrowFunctionExpression" &&
    getFunctionParameters(maybeArrowFunction).length === 0 &&
    maybeArrowFunction.body.type === "BlockStatement" &&
    maybeDepsArrary.type === "ArrayExpression" &&
    !args.some((arg) => hasComment(arg))
  );
}
```

Comment (Jaswanth-Sriram-Veturi) [reply src/language-js/print/call-arguments.js:329]: This looks great, I'll do these changes

**Review by sosukesuzuki — APPROVED** [review]

**Review by fisker — CHANGES_REQUESTED** [review]
We need limit the type of  the first argument. Maybe "Identifier" only?

**Comment by Jaswanth-Sriram-Veturi** [comment]
@fisker I'm not able to see your requested changes. Am I missing something? We are already checking if first argument type is "Identifier"

**Comment by fisker** [comment]
> We are already checking if first argument type is "Identifier"

Damn, missed..  Sorry.

File: src/language-js/print/call-arguments.js [hunk 2]
```diff
@@ -46,8 +46,17 @@ function printCallArguments(path, options, print) {
...
-    return ["(", print(["arguments", 0]), ", ", print(["arguments", 1]), ")"];
+    const parts = ["("];
+    for (let i = 0; i < args.length; i++) {
+      parts.push(print(["arguments", i]));
+      if (i !== args.length - 1) {
+        parts.push(", ");
+      }
+    }
+    parts.push(")");
+    return parts;
```
Comment (fisker) [inline src/language-js/print/call-arguments.js:59]: This part can simplify to `join(..., path.map(...))`, but better to use `iterateCallArgumentsPath(...)`.

Comment (Jaswanth-Sriram-Veturi) [reply src/language-js/print/call-arguments.js:59]:
```diff
@@ -49,12 +49,12 @@ function printCallArguments(path, options, print) {
   // useImperativeHandle(ref, () => { ... }, [foo, bar, baz])
   if (isReactHookCallWithDepsArray(args)) {
     const parts = ["("];
-    for (let i = 0; i < args.length; i++) {
-      parts.push(print(["arguments", i]));
-      if (i !== args.length - 1) {
+    iterateCallArgumentsPath(path, (arg, index) => {
+      parts.push(print(arg));
+      if (index !== args.length - 1) {
         parts.push(", ");
       }
-    }
+    });
     parts.push(")");
     return parts;
   }
~
``` 
This is how `iterateCallArgumentsPath` is used right?

**Review by fisker — APPROVED** [review]

**Pull request merged**